substrate/frame/contracts/src/exec.rs

```diff
@@ -1217,16 +1225,25 @@ where
 		value: BalanceOf<T>,
 		input_data: Vec<u8>,
 		allows_reentry: bool,
+		read_only: bool,
 	) -> Result<ExecReturnValue, ExecError> {
 		// Before pushing the new frame: Protect the caller contract against reentrancy attacks.
 		// It is important to do this before calling `allows_reentry` so that a direct recursion
 		// is caught by it.
 		self.top_frame_mut().allows_reentry = allows_reentry;
 
+		// Enable read-only access if requested; cannot disable it if already set.
+		let frame_read_only = read_only || self.top_frame().read_only;
+
 		let try_call = || {
 			if !self.allows_reentry(&to) {
 				return Err(<Error<T>>::ReentranceDenied.into())
 			}
+
+			// If the call value is non-zero and state change is not allowed, issue an error.
```

Comment:
let's explain  why here, instead of translating the if statement

Reply:
Ideally, it would be placed in `mod env`, but I would need to move most of the `Runtime::call` logic there. I already had another `read-only` checks in Stack (exec) for `call`, so I decided to keep all of them in one place.
I think now that we could move this logic to `Runtime::call` as checks for events disappeared from Stack.
WDYT?

Reply:
I think it should move to `mod env` so we have all the read only checks in one file. 

> but I would need to move most of the Runtime::call logic there.

I don't get this point. Can't you just move the check here and be done with it:
https://github.com/paritytech/polkadot-sdk/blob/4adfa37d14c0d81f09071687afb270ecdd5c2076/substrate/frame/contracts/src/wasm/runtime.rs#L844-L846

Reply:
I meant to move it to the `Runtime::call` function as you pointed out.